Python PETSc Example BP1

tylercollins5737 opened the pull request:
Hello!

This is a Python implementation of bpsraw.c for BP1 and was initially discussed in #1774 .  I'm aiming to recreate the benchmark behavior using libceed and petsc4py. I appreciate any feedback or indications I'm headed in the right direction!

It does run, but this is a work in progress. Remaining tasks:
- [x] fix the matrix operations to work like the MatShell C implementation
- [x] get PETSc Vec scatter working
- [x] QFunction 
- [x] Error function
- [x] add meaningful comments
- [x] improve the printed output metrics 
- [x] create a test

valeriabarra commented:
@tylercollins5737 , thank you for your work. Are you able to address the reviewer's comments?

tylercollins5737 commented on examples/python/bpsraw.py:
```diff
@@ -0,0 +1,155 @@
+    op.set_field("qdata", qdata_restriction, libceed.BASIS_NONE, qdata)
+    op.set_field("v", elem_restriction, basis, libceed.VECTOR_ACTIVE)
+
+    matrix = PETSc.Mat().createAIJ([num_dofs, num_dofs], comm=PETSc.COMM_WORLD)
+    matrix.setFromOptions()
+    matrix.setUp()
+    for i in range(num_dofs):
+        e_vec = ceed.Vector(num_dofs)
+        r_vec = ceed.Vector(num_dofs)
+        e_vec.set_value(0.0)
+        r_vec.set_value(0.0)
+        with e_vec.array_write() as arr:
+            arr[i] = 1.0
+        op.apply(e_vec, r_vec)
+        with r_vec.array_read() as r_arr:
+            matrix.setValues(range(num_dofs), [i], r_arr)
+    matrix.assemble()
```
Apologies for leaving this a bit stale. Addressing this led to a lot of other changes. I had trouble setting the matrix type to "shell" and applying context, kept getting segmentation faults. So, instead of using the "shell" type, I used the "python" type and set the context through a python class. Pulled that implementation from the petsc4py 2DPoisson example (about half way down) --> [link](https://petsc.org/release/petsc4py/petsc_python_types.html#petsc-python-matrix-type).  I'll tidy up what I have a submit an update soon.

jeremylt replied:
Overall this is looking like the right approach. The big thing is to try to avoid copying around data during frequent operations, like the matrix application

tylercollins5737 replied:
Thank you for all the suggestions! Definitely, keeping the data in place would mean less overhead and better efficiency

tylercollins5737 commented:
@valeriabarra I was able to get it working using the "python" matrix type, MatAIJ was not the best. I believe the "python" type and MatCreateShell can achieve the same thing.

jeremylt commented on examples/python/bpsraw.py:
```diff
@@ -0,0 +1,423 @@
+        self._xglob.fill(0.0)
+        self._xglob[lo:hi] = x.array_r
+        self.u.set_array(self._xglob,cmode=libceed.USE_POINTER)
+        self.v.set_value(0.0)
```
```suggestion
        self.v.set_value(0.0)

```
this shouldn't be needed

jeremylt commented on examples/python/bpsraw.py:
```diff
@@ -0,0 +1,423 @@
+        self.u.set_array(self._xglob,cmode=libceed.USE_POINTER)
+        self.v.set_value(0.0)
+        self.op.apply(self.u, self.v)
+        with self.v.array_read() as va:
+            y.setValues(range(lo, hi), va[lo:hi])
+        y.assemble()
```
Ideally, you'd set v to use y's array before you apply the libCEED operator

tylercollins5737 replied:
That makes more sense, thank you!

jeremylt commented on examples/python/bpsraw.py:
```diff
@@ -0,0 +1,423 @@
+                start += m_nodes[0] * m_nodes[1] * m_nodes[2]
+    return -1
+
+def CreateRestriction(ceed, mesh_elem, P, num_comp, ndofs, offsets):
+    return ceed.ElemRestriction(mesh_elem, P, num_comp, 1, ndofs, offsets, cmode=libceed.USE_POINTER)
```
If this just wraps another function call, I think its clearer to use the function call itself

tylercollins5737 replied:
Totally agree, I have another CreateRestriction function that reflects C but its a WIP

jeremylt commented:
Hi @tylercollins5737, I'm hoping to have this ready in case next semester's students want to build on this for BP 2-6. This is good work so far and will be very helpful to future users. Is there anything I can do to help here? I'm happy to help out on any pieces that have you stuck so we can get this moving again.

zatkins-dev commented:
@tylercollins5737 are you planning on coming back to this? I think there's a lot of good work here! If not, it may be time to close this PR.

tylercollins5737 commented:
> @tylercollins5737 are you planning on coming back to this? I think there's a lot of good work here! If not, it may be time to close this PR.

@zatkins-dev  Yes! I should have have a pretty complete update soon, ideally by the end of the weekend.

zatkins-dev commented:
Awesome! Happy to keep it open then, thanks!

tylercollins5737 committed:
python - add BP QFunctions for the Python examples

The gallery has Mass3DBuild and MassApply, but nothing like SetupMassRhs or Error, which depend on the exact solution the benchmark solves for. These follow the ex1-ex3 pattern: C sources next to the Python examples, built into the QFunctions library by qfunctions.c.

tylercollins5737 committed:
python - bpsraw.py, a Python port of the PETSc BP example

Port examples/petsc/bpsraw.c to Python with petsc4py and the libCEED Python bindings, solving BP1. A python PETSc Mat takes the place of the MatShell; its mult() runs the libCEED operator.

MatMult hands the PETSc Vec arrays straight to libCEED with CEED_USE_POINTER instead of copying them, on the host and on the device. petsc4py has no VecGetArrayAndMemType, so on the device the pointer comes from getCUDAHandle and is passed through the CUDA array interface.

tylercollins5737 committed:
python - test BP1 at degrees 1 to 3

The example picks its iteration limit from how long a first solve takes, so the tests pin it with ksp_max_it_clip to keep the iteration count the same every run, as the C test does.

tylercollins5737 commented:
- QFunctions are in the PR now (`qfunctions/bps/bp1.h`, `common.h`, plus `qfunctions.c`), using the `ex3_volume.py` approach you pointed me to. The gallery has `Mass3DBuild`/`MassApply` but nothing like `SetupMassRhs` or `Error`, so those have to be C QFunctions. 
- MatMult no longer copies data. The PETSc `Vec` arrays go to libCEED with `CEED_USE_POINTER`, on the host and on the device.
- Added tests (BP1, degrees 1–3) and matched the CLI to `bpsraw.c` (`-problem bp1`, `-ksp_max_it_clip`, `-write_solution`).
- Output matches `bpsraw.c` on CPU and GPU, across 1, 2 and 4 ranks and degrees 1–4. For example at 1 rank, degree 3: 16 iterations, max error 1.730043e-09, same as the C.

Two notes:
1. I copied `bp1.h`/`common.h` into `examples/python/qfunctions/bps/`, the way ex1–ex3 keep their own copies, so the example is self contained. 
2. `CeedVectorTakeArray` isn't wrapped in the Python bindings, so the borrowed arrays get rebound on the next MatMult instead of detached. Works as written, but not the same as the C.

jeremylt commented on examples/python/ex_test.py:
```diff
@@ -11,6 +11,7 @@
 import ex1_volume
 import ex2_surface
 import ex3_volume
+import bpsraw
```
since this file relies upon PETSc4Py, we'll want to test it separately from ex1-3, which only need libCEED

tylercollins5737 replied:
Sounds good, I'll create a separate test.py

jeremylt commented on examples/python/bpsraw.py:
```diff
@@ -0,0 +1,659 @@
+        # Apply libCEED operator
+        self.op_apply.apply(self.x_ceed, self.y_ceed)
+
+        # Restore arrays; the C also detaches them with CeedVectorTakeArray, which the Python bindings do not
```
We really should just add CeedVectorTakeArarry to the Python bindings - @zatkins-dev or I can hit that, though dunno how long it'll take to publish that

tylercollins5737 replied:
Just let me know about this. I can open a PR for it.

jeremylt replied:
https://github.com/CEED/libCEED/pull/1995

tylercollins5737 replied:
Awesome!

jeremylt commented on examples/python/bpsraw.py:
```diff
@@ -0,0 +1,659 @@
+                         if isinstance(value, int)}
+
+
+def Split3(size, reverse=False):
```
we should use the same function naming, style throughout. Also a brief docstring for each would be handy (though BPsRaw probably omitted that? If so, then no worries here)

tylercollins5737 replied:
Ah, yeah I forgot to rename them. I had them styled like C so I could spot differences easier. BPsRaw didn't really have any, but I'll fill them out.

jeremylt commented:
I have just a handful of comments, but this looks great!

tylercollins5737 committed:
python - test the BPs separately from ex1-ex3

ex1-ex3 only need libCEED, so moved the BP1 tests out of ex_test.py into their
own bps_test.py, which skips itself when petsc4py is not available. The test
target runs both.

tylercollins5737 committed:
python - use snake_case for the helper functions in bpsraw.py

Rename Split3, GlobalNodes, GlobalStart, CreateRestriction and ComputeErrorMax
to snake_case, matching the rest of the file and ex1-ex3.

tylercollins5737 committed:
python - add Args/Returns docstrings throughout bpsraw.py

Gave every function, class and method a docstring with Args and
Returns to match ex_common.py style.

tylercollins5737 commented:
So the BP tests skip when petsc4py isn't installed, and right now that's every CI setup because GitHub Actions has no PETSc, and GitLab CI has PETSc but not petsc4py. So they only really run locally at the moment.

jeremylt commented on examples/python/bps_test.py:
```diff
@@ -0,0 +1,67 @@
+        write_solution=False,
+        ksp_max_it_clip=[15, 15],
+    )
+    assert bpsraw.example_bps(args) == 0
```
Hmm, we either should use `assert` in the other test cases or not use it here. I *think* but am not sure that not having the `assert` results in this test just returning the error code of `example_bps`

tylercollins5737 replied:
Gotcha, I'll just remove them and match the current example tests

jeremylt commented on examples/python/bps_test.py:
```diff
@@ -0,0 +1,67 @@
+# -------------------------------------------------------------------------------
+
+
+def test_401(ceed_resource):
```
I'd pick a clearer set of test names, like `test_bp_1_degree_1` or such? Once this merges, I'll be adding an issue for a student to expand this to BP2, BP3, and BP4, so naming the tests by the BP will help us there

tylercollins5737 replied:
Sounds good! I think that naming convention makes perfect sense.

tylercollins5737 committed:
python - renamed BP tests by BP and degree; removed assert

tylercollins5737 committed:
python - detached borrowed arrays with take_array

Detached the arrays MatMult lends to libCEED with take_array from #1995,
matching CeedVectorTakeArray in the C.

tylercollins5737 commented:
1. assert: removed them, now matching ex_test.py.
2. names: renamed to `test_bp1_degree_{1,2,3}` — BP-keyed for the
   BPs expansion.
3. take_array: thanks for #1995! bpsraw.py now detaches with
   `take_array`, matching the C code.

jeremylt commented:
Ok, we need a cross check from someone who has petsc-py set up locally, but this looks awesome and ready to go as far as I can tell

maybe @zatkins-dev has petsc-py?

zatkins-dev commented:
I should be able to test locally today!

zatkins-dev commented:
Looks like the source branch needs to be rebased/merged from main -- it still doesn't have the take_array binding.

zatkins-dev commented:
> Ok, we need a cross check from someone who has petsc-py set up locally, but this looks awesome and ready to go as far as I can tell

But we're passing once I do that rebase locally, so I think this is good to go.

zatkins-dev commented:
@jeremylt we should be able to set up a test for this through CI too, I can draft an MR for it.

jeremylt commented:
Yeah, would be good to add, I just didn't have time for it

tylercollins5737 commented:
> Looks like the source branch needs to be rebased/merged from main -- it still doesn't have the take_array binding.

Yeah my bad, I had just added those lines locally. Rebased onto main!

zatkins-dev commented:
Looks good to me locally! I can add the test to CI in a follow-up.

zatkins-dev commented:
Alternatively, @tylercollins5737 can add this line to `.github/workflows/python-test-with-style.yml`

```diff
diff --git a/.github/workflows/python-test-with-style.yml b/.github/workflows/python-test-with-style.yml
index 5cf592c7..38db3e27 100644
--- a/.github/workflows/python-test-with-style.yml
+++ b/.github/workflows/python-test-with-style.yml
@@ -27,6 +27,7 @@ jobs:
       run: |
         pip install -r requirements.txt
         pip install -r requirements-test.txt
+        pip install -r requirements-test-petsc.txt
     - name: Python test
       env:
         CC: ${{ matrix.compiler }}
```

and add a new file `requirements-test-petsc.txt`:
```
mpi4py
petsc
petsc4py
```

tylercollins5737 commented:
> Alternatively, @tylercollins5737 can add this line to `.github/workflows/python-test-with-style.yml`
> 
> ```diff
> diff --git a/.github/workflows/python-test-with-style.yml b/.github/workflows/python-test-with-style.yml
> index 5cf592c7..38db3e27 100644
> --- a/.github/workflows/python-test-with-style.yml
> +++ b/.github/workflows/python-test-with-style.yml
> @@ -27,6 +27,7 @@ jobs:
>        run: |
>          pip install -r requirements.txt
>          pip install -r requirements-test.txt
> +        pip install -r requirements-test-petsc.txt
>      - name: Python test
>        env:
>          CC: ${{ matrix.compiler }}
> ```
> 
> and add a new file `requirements-test-petsc.txt`:
> 
> ```
> mpi4py
> petsc
> petsc4py
> ```

Cool, just pushed those changes.

zatkins-dev commented:
let me try to diagnose, this is a PETSc issue

zatkins-dev commented:
alright, you can drop that last commit and then rebase from main and it should work

tylercollins5737 commented:
Dropped the last commit and rebased to main.

zatkins-dev reviewed (approved):
Awesome, I think this looks great and we're passing tests. Assuming @jeremylt is on board, I'm good with merging!

jeremylt commented:
bedtime here, I'll do one last pass in the morning then merge!

tylercollins5737 commented:
Excellent news! Thank you both for the help and letting me work through this. Truly amazing what's been built in libCEED. Cheers!

jeremylt reviewed (approved):
ToDo: Create issue so to extend this to BP 2-4

jeremylt commented:
thanks for all the work here - this is a great add for users to see how to use libCEED in a Python based HPC project
